src/players/VAST.js

```diff
@@ -0,0 +1,211 @@
+import React, { Component } from 'react'
+import vast from 'vast-client'
+import { callPlayer } from '../utils'
+import createSinglePlayer from '../singlePlayer'
+import { FilePlayer } from './FilePlayer'
+
+const MATCH_URL = /^(https?:\/\/)?((bs\.serving-sys\.com)|(pubads\.g\.doubleclick\.net))/
+export class VAST extends Component {
+  static displayName = 'VAST';
+  static canPlay = url => MATCH_URL.test(url);
+
+  state = {
+    sources: [],
+    tracker: null
+  }
+
+  callPlayer = callPlayer;
+
+  createSourceFiles (mediaFiles = []) {
+    return mediaFiles.reduce((arr, {fileURL: src, mimeType: type}) => {
+      return [...arr, {src, type}]
+    }, [])
+  }
+
+  parseResponse (response) {
+    const {onEnded} = this.props
+    const {ads = []} = response
+
+    // find video creatives
+    // todo: handle companion ads
+    for (const ad of ads) {
+      const {creatives = []} = ad
+      for (const creative of creatives) {
+        const {mediaFiles = [], type} = creative
+        if (type === 'linear') {
+          const sources = this.createSourceFiles(mediaFiles)
+          if (sources.length) {
+            return this.setState({
+              sources,
+              tracker: new vast.tracker(ad, creative)
+            })
+          }
+        }
+      }
+
+      // Inform ad server we can't find suitable media file for this ad
+      vast.util.track(ad.errorURLTemplates, { ERRORCODE: 403 })
+      onEnded()
+    }
+  }
+
+  load (url) {
+    vast.client.get(url, { withCredentials: true }, (response, error) => {
+      if (error) {
+        return this.props.onError(error)
+      }
+      this.parseResponse(response)
+      const {tracker} = this.state
+      if (tracker) {
+        tracker.on('clickthrough', this.openAdLink)
+      }
+    })
+  }
+
+  // todo: add skip functionality
+  skip () {
+    const {props: {onEnded}, state: {tracker}} = this
```

Comment:
I personally don't like this style if destructuring, but that is a personal thing.